dnf/transaction_sr.py

```diff
@@ -0,0 +1,600 @@
+                elif action == "Upgrade":
+                    self._swdb_environment_upgrade(env_id, pkg_types, env_data["groups"])
+                elif action == "Removed":
+                    self._swdb_environment_remove(env_id, pkg_types, env_data["groups"])
+                else:
+                    errors.append(TransactionError(
+                        _('Unexpected value of environment action "{action}" for environment "{env}".')
+                            .format(action=action, env=env_id)
+                    ))
+            except KeyError as e:
+                errors.append(TransactionError(
+                    _('Missing object key "{key}" in an environment.').format(key=e.args[0])
+                ))
+            except TransactionError as e:
+                errors.append(e)
+
+        if errors:
+            raise TransactionFileError(fn, errors)
+
+    def post_transaction(self):
+        """
+        Sets reasons in the transaction history to values from the stored transaction.
+
+        Also serves to check whether additional packages were pulled in by the
+        transaction, which results in an error (unless ignore_extras is True).
+        """
+
+        if not self._base.transaction:
+            return
+
+        errors = []
+
+        for tsi in self._base.transaction:
+            try:
+                pkg = tsi.pkg
+            except KeyError as e:
+                # the transaction item has no package, happens for action == "Reason Change"
+                continue
+
+            nevra = str(pkg)
+
+            if nevra not in self._nevra_cache:
+                # if ignore_installed is True, we don't want to check for
+                # Upgraded/Downgraded/Reinstalled extras in the transaction,
+                # basically those may be installed and we are ignoring them
+                if not self._ignore_installed or not tsi.action in (
+                    libdnf.transaction.TransactionItemAction_UPGRADED,
+                    libdnf.transaction.TransactionItemAction_DOWNGRADED,
+                    libdnf.transaction.TransactionItemAction_REINSTALLED
+                ):
+                    msg = _('Package nevra "{nevra}", which is not present in the transaction file, was pulled '
+                        'into the transaction.'
+                    ).format(nevra=nevra)
+
+                    if not self._ignore_extras:
+                        errors.append(TransactionError(msg))
+                    else:
+                        self._warnings.append(msg)
+
+            try:
+                replay_reason = self._nevra_reason_cache[nevra]
+
+                if tsi.action in (
+                    libdnf.transaction.TransactionItemAction_INSTALL,
+                    libdnf.transaction.TransactionItemAction_REMOVE
+                ) or libdnf.transaction.TransactionItemReasonCompare(replay_reason, tsi.reason) > 0:
+                    tsi.reason = replay_reason
+            except KeyError as e:
```

Comment:
I suggest that `if nevra not in self._nevra_cache:` already did dte same job, probably both parts could be merged. What do you  think? This only minor comment and does't have any impact on functionality.

Reply:
`if nevra not in self._nevra_cache:` is a different container, here the error happens in `self._nevra_reason_cache`. I could unify the style to do an `in` check instead of catching the exception, but it's really a minor difference...